{
  "classifiers": [
    "development status :: 5 - production/stable",
    "intended audience :: developers",
    "intended audience :: system administrators",
    "license :: osi approved :: apache software license",
    "operating system :: os independent",
    "programming language :: python :: 3",
    "programming language :: python :: 3.10",
    "programming language :: python :: 3.11",
    "programming language :: python :: 3.7",
    "programming language :: python :: 3.8",
    "programming language :: python :: 3.9",
    "topic :: utilities"
  ],
  "description": "# aws-error-utils\n**making botocore.exceptions.clienterror easier to deal with**\n\nall aws service exceptions are raised by `boto3` as a `botocore.exceptions.clienterror`, with the contents of the exception indicating what kind of exception happened.\nthis is not very pythonic, and the contents themselves are rather opaque, most being held in dicts rather than as properties.\nthe functions in this package help dealing with that, to make your code less verbose and require less memorization of `clienterror` contents.\n\n## installation\n\n[the package is on pypi](https://pypi.org/project/aws-error-utils/) for pip-installing, but you can also just copy the [`aws_error_utils.py` file](https://raw.githubusercontent.com/benkehoe/aws-error-utils/stable/aws_error_utils/aws_error_utils.py) into your project (note that the standalone file requires python 3.7 or greater).\n\n## usage\nif you've got code like this:\n\n```python\nimport boto3, botocore.exceptions\n\ns3 = boto3.client('s3')\ntry:\n    s3.get_object(bucket='my-bucket', key='example')\nexcept botocore.exceptions.clienterror as error:\n    if error.response['error']['code'] == 'nosuchbucket':\n        print(error.response['error']['message'])\n        # error handling\n    else:\n        raise\n```\n\nyou can replace it with:\n\n```python\nimport boto3\nfrom aws_error_utils import errors\n\ns3 = boto3.client('s3')\ntry:\n    s3.get_object(bucket='my-bucket', key='example')\nexcept errors.nosuchbucket as error:\n# or\n# except catch_aws_error('nosuchbucket') as error:\n    print(error.message)\n    # error handling\n```\n\nif you have trouble remembering where all the contents in `clienterror` are, like these:\n\n```python\nclient_error.response['error']['code']\nclient_error.response['error']['message']\nclient_error.response['responsemetadata']['httpstatuscode']\nclient_error.operation_name\n```\n\nyou can replace it with:\n\n```python\nimport boto3\nfrom aws_error_utils import get_aws_error_info\n\nerr_info = get_aws_error_info(client_error)\n\nerr_info.code\nerr_info.message\nerr_info.http_status_code\nerr_info.operation_name\n```\n\nif you're using `errors` or `catch_aws_error()`, you can skip the `get_aws_error_info()` step, because the fields are set directly on the `clienterror` object:\n\n```python\nimport boto3\nfrom aws_error_utils import errors\n\ns3 = boto3.client('s3')\ntry:\n    s3.get_object(bucket='my-bucket', key='example')\nexcept errors.nosuchbucket as error:\n    error.code\n    error.message\n    error.http_status_code\n    error.operation_name\n```\n\nadditionally, when you *do* need to use `clienterror`, it's imported by `aws_error_utils` so you can just use `aws_error_utils.clienterror` rather than `botocore.exceptions.clienterror` (the same is true for `botocoreerror`, the base class of all non-`clienterror` exceptions).\n\n## `errors`\nit's easiest to use the `errors` class if you don't have complex conditions to match.\nusing the error code as a field name in an `except` block will match that error code.\nadditionally, when you use this style, it sets the fields from `awserrorinfo` (see below) directly on the `clienterror` object.\nfor example:\n\n```python\nimport boto3\nfrom aws_error_utils import errors\n\ns3 = boto3.client('s3')\ntry:\n    s3.get_object(bucket='my-bucket', key='example')\nexcept errors.nosuchbucket as error:\n    print(error.message)\n\n    # error handling\n```\n\nyou can include multiple error codes in an `except` statement, though this is slower than combining them with a single `catch_aws_error()` call.\n\n```python\nimport boto3\nfrom aws_error_utils import errors\n\ns3 = boto3.client('s3')\ntry:\n    s3.get_object(bucket='my-bucket', key='example')\nexcept (errors.nosuchbucket, errors.nosuchkey) as error:\n    print(error.message)\n\n    # error handling\n```\n\nyou can catch all `clienterror`s with `errors.all`.\n\nyou can only use this style for error codes that work as python property names.\nfor error codes like ec2's `invalidinstanceid.notfound`, you have to use `catch_aws_error()` (see below).\n\nunfortunately, you cannot get tab completion for error codes on the `errors` class, as a comprehensive list of error codes is not available as a python package (`botocore` has a small number, but they are few and incomplete).\n\nnote that the value of `errors.nosuchbucket` is not an exception class representing the `nosuchbucket` error, it is an alias for `catch_aws_error('nosuchbucket')`.\nit can only be used in an `except` statement; it will raise `runtimeerror` otherwise.\nyou also cannot instantiate the `errors` class.\n\n## `catch_aws_error()`\nthe function takes as input error code(s), and optionally operation name(s), to match against the current raised exception. if the exception matches, the `except` block is executed.\nif your error handling still needs the error object, you can still use an `as` expression, otherwise it can be omitted (just `except catch_aws_error(...):`).\nadditionally, `catch_aws_error()` sets the fields from `awserrorinfo` (see below) directly on the `clienterror` object.\n\n```python\nimport boto3\nfrom aws_error_utils import catch_aws_error\n\ns3 = boto3.client('s3')\ntry:\n    s3.get_object(bucket='my-bucket', key='example')\nexcept catch_aws_error('nosuchbucket') as error:\n    print(error.message)\n\n    # error handling\n```\n\nyou can provide error codes either as positional args, or as the `code` keyword argument with either as a single string or a list of strings.\n\n```python\ncatch_aws_error('nosuchbucket')\ncatch_aws_error(code='nosuchbucket')\n\ncatch_aws_error('nosuchbucket', 'nosuchkey')\ncatch_aws_error(code=['nosuchbucket', 'nosuchkey'])\n```\n\nif there are multiple api calls in the `try` block, and you want to match against specific ones, the `operation_name` keyword argument can help.\nsimilar to the `code` keyword argument, the operation name(s) can be provided as either as a single string or a list of strings.\n\n```python\nimport boto3\nfrom aws_error_utils import catch_aws_error\n\ntry:\n    s3 = boto3.client('s3')\n    s3.list_objects_v2(bucket='bucket-1')\n    s3.get_object(bucket='bucket-2', key='example')\nexcept catch_aws_error('nosuchbucket', operation_name='getobject') as error:\n    # this will be executed if the getobject operation raises nosuchbucket\n    # but not if the listobjects operation raises it\n```\n\nyou must provide an error code.\nto match exclusively against operation name, use the `aws_error_utils.all_codes` token.\nfor completeness, there is also an `all_operations` token.\n\n```python\nimport boto3\nfrom aws_error_utils import catch_aws_error\n\ntry:\n    s3 = boto3.client('s3')\n    s3.list_objects_v2(bucket='bucket-1')\n    s3.get_object(bucket='bucket-1', key='example')\nexcept catch_aws_error(all_codes, operation_name='listobjectsv2') as e:\n    # this will execute for all clienterror exceptions raised by the listobjectsv2 call\n```\n\nfor more complex conditions, instead of providing error codes and operation names, you can provide a callable to evaluate the exception.\nnote that unlike error codes, you can only provide a single callable.\n\n```python\nimport re\nimport boto3\nfrom aws_error_utils import catch_aws_error, get_aws_error_info\n\ndef matcher(e):\n    info = get_aws_error_info(e)\n    return re.search('does not exist', info.message)\n\ntry:\n    s3 = boto3.client('s3')\n    s3.list_objects_v2(bucket='bucket-1')\nexcept catch_aws_error(matcher) as e:\n    # this will be executed if e is a clienterror and matcher(e) returns true\n    # note the callable can assume the exception is a clienterror\n```\n\n## `get_aws_error_info()`\nthis function takes a returns an `awserrorinfo` object, which is a `collections.namedtuple` with the following fields:\n\n* `code`\n* `message`\n* `http_status_code`\n* `operation_name`\n* `response` (the raw response dictionary)\n\nif you're not modifying your `except` statements to use `catch_aws_error()`, this function can be useful instead of remembering exactly how this information is stored in the `clienterror` object.\n\nif you're using `catch_aws_error()`, this function isn't necessary, because it sets these fields directly on the `clienterror` object.\n\n## `aws_error_matches()`\nthis is the matching logic behind `catch_aws_error()`.\nit takes a `clienterror`, with the rest of the arguments being error codes and operation names identical to `catch_aws_error()`, except that it does not support providing a callable.\n\n## `make_aws_error()`\nfor testing, you can create `clienterror`s with the `make_aws_error()` function, which takes the following arguments:\n* `code` (required)\n* `message` (required)\n* `operation_name` (required)\n* `http_status_code`\n* `response`\n\nif a `response` dict is provided, it will be copied rather than modified.\n\n",
  "docs_url": null,
  "keywords": "",
  "license": "apache-2.0",
  "name": "aws-error-utils",
  "package_url": "https://pypi.org/project/aws-error-utils/",
  "project_url": "https://pypi.org/project/aws-error-utils/",
  "project_urls": {
    "Homepage": "https://github.com/benkehoe/aws-error-utils",
    "Repository": "https://github.com/benkehoe/aws-error-utils"
  },
  "release_url": "https://pypi.org/project/aws-error-utils/2.7.0/",
  "requires_dist": [
    "botocore"
  ],
  "requires_python": ">=3.7,<4",
  "summary": "error-handling functions for boto3/botocore",
  "version": "2.7.0",
  "releases": [],
  "developers": [
    "ben_kehoe"
  ],
  "kwds": "catch_aws_error aws_error_utils get_aws_error_info make_aws_error aws_error_matches",
  "license_kwds": "apache-2.0",
  "libtype": "pypi",
  "id": "pypi_aws_error_utils",
  "homepage": "https://github.com/benkehoe/aws-error-utils",
  "release_count": 11,
  "dependency_ids": [
    "pypi_botocore"
  ]
}